app/src/main/java/echoserver/SocketIo.java

```diff
@@ -0,0 +1,40 @@
+package echoserver;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.net.Socket;
+
+public class SocketIo {
+  BufferedReader in;
+  PrintWriter out;
+
+  public SocketIo() {
+    this.in = null;
+    this.out = null;
+  }
+
+  public BufferedReader createSocketInput(Socket clientSocket) throws IOException {
+    in = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
+    return in;
+  }
+
+  public PrintWriter createSocketOutput(Socket clientSocket) throws IOException {
+    out = new PrintWriter(clientSocket.getOutputStream(), true);
+    return out;
+  }
+
+  public void send(String message) {
+    out.println(message);
```

Comment:
theres a potential runtime error here where if you create an instance of `SocketIo` but not call `createSocketOuput` this method would fail. we should find a way to avoid that. ill leave the decision to you :D

Reply:
Updated to create the I/O stream reader/writer in the SocketIo constructor method to avoid this error